### PR TITLE
build the docs in python 3.9

### PR DESCRIPTION
**Pull request by 1ucian0**
Currently, docs building is failing on Python 3.9. This PR fixes the issue (by importing numpy in qiskit/primitives/estimator_result.py and updates azure to build using Python 3.9.

Thanks @jakelishman for the help debugging this!

### CODE REVIEW

**Review by jakelishman — COMMENTED** [review]
In general this is fine for me.  For future reference: `TYPE_CHECKING` does not work in Sphinx builds, because Sphinx isn't a static type-checker, so it (correctly) doesn't set that variable to `True`.  This means that Sphinx will never be able to resolve type hints that would need a true cyclic import to handle (without some funky `from __future__ import annotations` and end-of-file trailing `import`s).  

The `typing` module in Python 3.9 got some upgrades that means it's somewhat more correct about erroring out in these situations, whereas in previous Python versions, the type-hint resolution would silently fail, and often produce documentation that built without error, but was actually showing something invalid.  In this particular case, the documentation ended up looking mostly ok, but only because Sphinx resorted to just displaying the text verbatim.  In most forward-reference situations in Python before 3.9, we'd actually just get the literal text `ForwardRef` appear instead.  I think which behaviour you get depends a little on whether `from __future__ import annotations` is defined, but either way, Sphinx _requires_ that type hints can be resolved at run-time (a larger restriction than Mypy, which uses static analysis).

**Comment by jakelishman** [comment]
Python 3.9 is (mostly) just a stricter version, so 3.8 should build anything that 3.9 builds.  That said, yeah, we should bump the metapackage version too once this is passing.

The linter failure might need some thought - it's possible that while we support Python 3.7 and 3.8, we just might not want to use class-subscript typing (e.g. `np.ndarray[...]`).  With `from __future__ import annotations`, we can probably make it work if we up the version of Python for the linter, but that somewhat defeats the point, since we want the lint to pass on all versions of Python we support.  It's _possible_ that it's just an oversight in older versions of pylint causing the failure, but it's probably more likely differences to how `type.__getitem__` got handled in newer versions of Python.

(This is a lot of why I don't like Python's implementation of type hints...)

**Comment by ikkoham** [comment]
`set_type_checking_flag` may be helpful to build the docs.

BTW, why do we use this version of `sphinx` and `sphinx-autodoc-typehint`? It might be better to upgrade the version.

**Comment by jakelishman** [comment]
We use the latest version of Sphinx (that took me quite a long time to get working across all the modules), but I don't remember why we pin `sphinx-autodoc-typehints`.  I didn't know about the `set_type_checking_flag` setting, thanks. It won't solve all our problems - things like circular imports will still fail, because Sphinx can't resolve them statically since it's actually doing things at runtime - but it might help in situations like this.

File: azure-pipelines.yml [hunk 1]
```diff
@@ -59,6 +59,10 @@ parameters:
     type: string
     default: "3.8"
 
+  - name: "documentationPythonVersion"
+    displayName: "Version of Python to use to build Sphinx documentation"
+    type: string
+    default: "3.9"
```
[the remark this answers is not in the document]
Comment (1ucian0) [reply azure-pipelines.yml:65]: done in 5eada5f0f9ca370ec807b6baeba6e9fcad6582d2

**Comment by manoelmarques** [comment]
are there any thoughts regarding sphinx-panels that is basically a dead project and pins Sphinx at version < 5 ?

**Comment by jakelishman** [comment]
It was added as a dependency ages ago (it's used in the main metapackage documentation, if I remember correctly, and possibly the transpiler page).  [According to its repository](https://github.com/executablebooks/sphinx-panels), it's staled and we should swap to a new package called [`sphinx-design`](https://github.com/executablebooks/sphinx-design).

We should be able to do that across Qiskit without too much trouble, since we don't use it heavily.

**Comment by 1ucian0** [comment]
Moving to `sphinx-design` would also help to fix https://github.com/Qiskit/qiskit/issues/1501

**Comment by HuangJunye** [comment]
Should we create an issue to migrate `sphinx-panels` to `sphinx-design` and another one for upgrading to sphinx 5.0?

**Comment by 1ucian0** [comment]
> Should we create an issue to migrate `sphinx-panels` to `sphinx-design` and another one for upgrading to sphinx 5.0?

It seems to me that we can get most of the benefits by switching to `sphinx-design` while on sphinx 4, at least for now. What do you think @jakelishman @mtreinish ?

**Comment by jakelishman** [comment]
An issue for `sphinx-design` is fine.  I don't think we manually pin Sphinx to the 4.x series, so if we remove the blockers, CI should just start pulling in Sphinx 5 anyway.  If that causes problems, we can start pinning Sphinx to a given major release.

File: requirements-dev.txt [hunk 2]
```diff
@@ -19,7 +19,7 @@ reno>=3.4.0
 Sphinx>=3.0.0
 qiskit-sphinx-theme>=1.6
 qiskit-toqm>=0.0.4;platform_machine != 'aarch64' or platform_system != 'Linux'
-sphinx-autodoc-typehints<1.14.0
+sphinx-autodoc-typehints~=1.18
```
Comment (jakelishman) [inline requirements-dev.txt:22]: This requirement is actually just `>=1,<2` - I don't know if you were intending to do something else (like `>=1.18`, perhaps?).

**Review by jakelishman — APPROVED** [review]
The one requirement comment aside (doesn't necessarily need changing), this looks good to me. Let's move to get it merged today, so we don't introduce any regressions later.

**Pull request merged**